File: python/decidekit/providers/http.py
from __future__ import annotations

import json
import time
from typing import Any, Callable
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from ..errors import ProviderError
from ..types import DecisionQuestion, JsonValue, ProviderResponse, Usage

Transport = Callable[[Request, float], Any]


class HttpDecisionProvider:
    def __init__(
        self,
        *,
        name: str,
        endpoint: str,
        api_key: str,
        default_model: str,
        headers: dict[str, str] | None = None,
        max_retries: int = 2,
        timeout: float = 30.0,
        input_cost_per_million: float | None = None,
        output_cost_per_million: float | None = None,
        transport: Transport | None = None,
        transform_body: Callable[[JsonValue, dict[str, DecisionQuestion], str], dict[str, Any]] | None = None,
    ):
        if not api_key:
            raise ProviderError(f"{name} API key is required")
        self.name = name
        self.endpoint = endpoint
        self.api_key = api_key
        self.default_model = default_model
        self.headers = headers or {}
        self.max_retries = max_retries
        self.timeout = timeout
        self.input_cost_per_million = input_cost_per_million
        self.output_cost_per_million = output_cost_per_million
        # urlopen's second positional argument is request data, not timeout.
        # Keep the injected transport contract as (request, timeout) while
        # adapting the standard-library function correctly.
        self.transport = transport or (lambda request, timeout: urlopen(request, timeout=timeout))
        self.transform_body = transform_body

    def evaluate(
        self,
        state: JsonValue,
        questions: dict[str, DecisionQuestion],
        model: str | None = None,
    ) -> ProviderResponse:
        effective_model = model or self.default_model
        body = (
            self.transform_body(state, questions, effective_model)
            if self.transform_body
            else {"state": state, "questions": {key: value.to_payload() for key, value in questions.items()}, "model": effective_model}
        )
        request = Request(
            self.endpoint,
            data=json.dumps(body).encode("utf-8"),
            method="POST",
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json",
                **self.headers,
            },
        )
        started = time.perf_counter()
        last_error: Exception | None = None
        for attempt in range(self.max_retries + 1):
            try:
                with self.transport(request, self.timeout) as response:
                    raw = response.read().decode("utf-8")
                parsed = json.loads(raw)
                answers = parsed.get("answers")
                if not isinstance(answers, dict):
                    raise ProviderError(f"{self.name} returned a response without an answers object")
                usage_data = parsed.get("usage") or {}
                reported_cost = usage_data.get("costUsd", usage_data.get("cost", usage_data.get("total_cost")))
                if reported_cost is None:
                    reported_cost = parsed.get("costUsd", parsed.get("cost", parsed.get("total_cost")))
                cost_usd = float(reported_cost) if reported_cost is not None else None
                if cost_usd is None and self.input_cost_per_million is not None and self.output_cost_per_million is not None:
                    cost_usd = (
                        int(usage_data.get("input_tokens", usage_data.get("inputTokens", 0))) * self.input_cost_per_million
                        + int(usage_data.get("output_tokens", usage_data.get("outputTokens", 0))) * self.output_cost_per_million
                    ) / 1_000_000
                usage = Usage(
                    int(usage_data.get("input_tokens", usage_data.get("inputTokens", 0))),
                    int(usage_data.get("output_tokens", usage_data.get("outputTokens", 0))),
                    cost_usd,
                )
                return ProviderResponse(
                    answers=answers,
                    provider=self.name,
                    model=str(parsed.get("model", effective_model)),
                    usage=usage,
                    latency_ms=round((time.perf_counter() - started) * 1000),
                )
            except HTTPError as error:
                response_body = error.read().decode("utf-8", errors="replace")
                last_error = ProviderError(
                    f"{self.name} request failed with HTTP {error.code}: {response_body[:500]}",
                    error.code,
                    response_body,
                )
                if error.code not in {429, 500, 502, 503, 504} or attempt >= self.max_retries:
                    raise last_error from error
            except (URLError, TimeoutError, json.JSONDecodeError) as error:
                last_error = ProviderError(f"{self.name} request failed: {error}")
                if attempt >= self.max_retries:
                    raise last_error from error
            if last_error is not None:
                time.sleep(0.2 * (2**attempt))
        raise last_error or ProviderError(f"{self.name} request failed")

File: src/providers/http.ts
import { ProviderError } from "../errors.js";
import type { DecisionAnswer, DecisionProvider, DecisionRequest, ProviderResponse, Usage } from "../types.js";

export type FetchLike = typeof fetch;

interface RawProviderPayload {
  answers?: Record<string, DecisionAnswer>;
  usage?: {
    input_tokens?: number;
    output_tokens?: number;
    inputTokens?: number;
    outputTokens?: number;
    cost?: number;
    total_cost?: number;
    costUsd?: number;
  };
  cost?: number;
  total_cost?: number;
  costUsd?: number;
  model?: string;
}

export interface HttpProviderOptions {
  name: string;
  endpoint: string;
  apiKey: string;
  defaultModel: string;
  headers?: Record<string, string>;
  fetch?: FetchLike;
  maxRetries?: number;
  timeoutMs?: number;
  pricing?: {
    inputCostPerMillion: number;
    outputCostPerMillion: number;
  };
  transformBody?: (request: DecisionRequest, model: string) => unknown;
  transformResponse?: (body: unknown) => RawProviderPayload;
}

function parseUsage(
  usage: RawProviderPayload["usage"],
  payload: RawProviderPayload,
  pricing: HttpProviderOptions["pricing"],
): Usage | undefined {
  if (!usage) return undefined;
  const reportedCost =
    usage.costUsd ?? usage.cost ?? usage.total_cost ?? payload.costUsd ?? payload.cost ?? payload.total_cost;
  const costUsd =
    typeof reportedCost === "number"
      ? reportedCost
      : pricing
        ? ((usage.input_tokens ?? usage.inputTokens ?? 0) * pricing.inputCostPerMillion +
            (usage.output_tokens ?? usage.outputTokens ?? 0) * pricing.outputCostPerMillion) /
          1_000_000
        : undefined;
  return {
    inputTokens: usage.input_tokens ?? usage.inputTokens ?? 0,
    outputTokens: usage.output_tokens ?? usage.outputTokens ?? 0,
    ...(costUsd === undefined ? {} : { costUsd }),
  };
}

function sleep(milliseconds: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(resolve, milliseconds);
    signal?.addEventListener(
      "abort",
      () => {
        clearTimeout(timer);
        reject(signal.reason ?? new Error("Request aborted"));
      },
      { once: true },
    );
  });
}

function combineSignals(signal: AbortSignal | undefined, timeoutMs: number): AbortSignal {
  return signal ? AbortSignal.any([signal, AbortSignal.timeout(timeoutMs)]) : AbortSignal.timeout(timeoutMs);
}

export class HttpDecisionProvider implements DecisionProvider {
  readonly name: string;
  readonly #options: HttpProviderOptions;
  readonly #fetch: FetchLike;

  constructor(options: HttpProviderOptions) {
    if (!options.apiKey) throw new ProviderError(`${options.name} API key is required.`);
    this.name = options.name;
    this.#options = options;
    this.#fetch = options.fetch ?? globalThis.fetch;
  }

  async evaluate(request: DecisionRequest): Promise<ProviderResponse> {
    const model = request.model ?? this.#options.defaultModel;
    const startedAt = performance.now();
    const maxRetries = this.#options.maxRetries ?? 2;
    let lastError: unknown;

    for (let attempt = 0; attempt <= maxRetries; attempt += 1) {
      try {
        const response = await this.#fetch(this.#options.endpoint, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${this.#options.apiKey}`,
            "Content-Type": "application/json",
            ...this.#options.headers,
          },
          body: JSON.stringify(
            this.#options.transformBody?.(request, model) ?? {
              state: request.state,
              questions: request.questions,
              model,
            },
          ),
          signal: combineSignals(request.signal, this.#options.timeoutMs ?? 30_000),
        });

        if (!response.ok) {
          const responseBody = await response.text().catch(() => "");
          const error = new ProviderError(
            `${this.name} request failed with HTTP ${response.status}${responseBody ? `: ${responseBody.slice(0, 500)}` : ""}`,
            response.status,
            responseBody,
          );
          if ((response.status === 429 || response.status >= 500) && attempt < maxRetries) {
            lastError = error;
            await sleep(200 * 2 ** attempt, request.signal);
            continue;
          }
          throw error;
        }

        const rawBody: unknown = await response.json();
        const body = this.#options.transformResponse?.(rawBody) ?? (rawBody as RawProviderPayload);
        if (!body.answers || typeof body.answers !== "object") {
          throw new ProviderError(`${this.name} returned a response without an answers object.`);
        }
        const usage = parseUsage(body.usage, body, this.#options.pricing);
        return {
          answers: body.answers,
          provider: this.name,
          model: body.model ?? model,
          ...(usage ? { usage } : {}),
          latencyMs: Math.round(performance.now() - startedAt),
        };
      } catch (error) {
        lastError = error;
        if (error instanceof ProviderError || attempt >= maxRetries) throw error;
        await sleep(200 * 2 ** attempt, request.signal);
      }
    }

    throw lastError instanceof Error ? lastError : new ProviderError(`${this.name} request failed.`);
  }
}
